Source/DafnyCore/Util.cs

```diff
@@ -231,7 +313,13 @@ public static class Util {
             case @"\r": yield return '\r'; break;
             case @"\t": yield return '\t'; break;
             case { } when s.StartsWith(@"\u"):
-              yield return (char)Convert.ToInt32(s[2..], 16);
+              yield return Convert.ToInt32(s[2..], 16);
+              break;
+            case { } when s.StartsWith(@"\U"):
+              yield return Convert.ToInt32(s[3..^1], 16);
```

Comment:
That's right, but the precondition documented on the function by "the well-parsed string p" implies we can assume it, since the parser invokes `ValidateEscaping`. 

It's a bit tempting to introduce a mini AST-like data structure for string content with escapes, to introduce a bit more type safety into this approach...